File: app/src/main/java/com/sam_chordas/android/stockhawk/service/StockWidgetUpdateIntentService.java
package com.sam_chordas.android.stockhawk.service;

import android.app.IntentService;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.widget.RemoteViews;

import com.sam_chordas.android.stockhawk.model.Quote;

import org.jetbrains.annotations.Nullable;

import kotlin.Metadata;
import kotlin.jvm.internal.IntCompanionObject;
import kotlin.jvm.internal.Intrinsics;

@Metadata(
   mv = {1, 1, 1},
   bv = {1, 0, 0},
   k = 1,
   d1 = {"\u0000(\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0002\b\u0004\n\u0002\u0010\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0010\b\n\u0000\u0018\u00002\u00020\u0001B\u0005¢\u0006\u0002\u0010\u0002J\u0012\u0010\b\u001a\u00020\t2\b\u0010\n\u001a\u0004\u0018\u00010\u000bH\u0014J\u0010\u0010\f\u001a\u00020\t2\u0006\u0010\r\u001a\u00020\u000eH\u0002R\u0019\u0010\u0003\u001a\n \u0005*\u0004\u0018\u00010\u00040\u0004¢\u0006\b\n\u0000\u001a\u0004\b\u0006\u0010\u0007¨\u0006\u000f"},
   d2 = {"Lcom/sam_chordas/android/stockhawk/service/StockWidgetUpdateIntentService;", "Landroid/app/IntentService;", "()V", "TAG", "", "kotlin.jvm.PlatformType", "getTAG", "()Ljava/lang/String;", "onHandleIntent", "", "intent", "Landroid/content/Intent;", "updateWidget", "widgetId", "", "production sources for module app"}
)
public final class StockWidgetUpdateIntentService extends IntentService {
   private final String TAG = StockWidgetUpdateIntentService.class.getSimpleName();

   public final String getTAG() {
      return this.TAG;
   }

   protected void onHandleIntent(@Nullable Intent intent) {
      if(intent != null) {
         String var2 = intent.getAction();
         if(Intrinsics.areEqual(var2, StockWidgetUpdateIntentServiceKt.getACTION_UPDATE_WIDGETS())) {
            int[] var5 = intent.getIntArrayExtra(StockWidgetUpdateIntentServiceKt.getEXTRA_WIDGET_IDS());

            for(int var4 = 0; var4 < var5.length; ++var4) {
               int appWidgetId = var5[var4];
               this.updateWidget(appWidgetId);
            }
         } else if(Intrinsics.areEqual(var2, StockWidgetUpdateIntentServiceKt.getACTION_UPDATE_WIDGET())) {
            this.updateWidget(intent.getIntExtra(StockWidgetUpdateIntentServiceKt.getEXTRA_WIDGET_ID(), IntCompanionObject.MIN_VALUE));
         } else {
            Log.wtf(this.TAG, "Unexpected action: " + intent.getAction());
         }
      }

   }

   private final void updateWidget(int widgetId) {
      Quote var10000 = StockWidgetUpdateIntentServiceKt.loadQuote((Context)this, widgetId);
      if(var10000 != null) {
         Quote quote = var10000;
         RemoteViews views = new RemoteViews(this.getPackageName(), 2130968655);
         views.setTextViewText(2131624080, (CharSequence)quote.symbol);

         try {
            String bg = quote.change;
            float e = Float.parseFloat(bg);
            int bg1 = e < (float)0?2130837652:2130837651;
            views.setInt(2131624119, "setBackgroundResource", bg1);
         } catch (NumberFormatException var6) {
            Log.wtf(this.TAG, (Throwable)var6);
         }

         views.setTextViewText(2131624081, (CharSequence)quote.bidPrice);
         views.setTextViewText(2131624082, (CharSequence)quote.change);
         AppWidgetManager.getInstance((Context)this).updateAppWidget(widgetId, views);
      }
   }

   public StockWidgetUpdateIntentService() {
      super("StockWidgetUpdateIntentService");
      this.setIntentRedelivery(true);
   }
}

File: app/src/main/java/com/sam_chordas/android/stockhawk/service/StockWidgetUpdateIntentService.kt
package com.sam_chordas.android.stockhawk.service

import android.app.IntentService
import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.util.Log
import android.widget.RemoteViews
import com.sam_chordas.android.stockhawk.R
import com.sam_chordas.android.stockhawk.data.QuoteColumns
import com.sam_chordas.android.stockhawk.data.QuoteProvider
import com.sam_chordas.android.stockhawk.model.Quote
import com.sam_chordas.android.stockhawk.ui.QUOTE_LIST_PROJECTION

class StockWidgetUpdateIntentService : IntentService("StockWidgetUpdateIntentService") {
    val TAG = StockWidgetUpdateIntentService::class.java.simpleName

    init {
        setIntentRedelivery(true)
    }

    override fun onHandleIntent(intent: Intent?) {
        if (intent != null) when (intent.action) {
            ACTION_UPDATE_WIDGETS -> {
                for (appWidgetId in intent.getIntArrayExtra(EXTRA_WIDGET_IDS)) {
                    updateWidget(appWidgetId)
                }
            }
            ACTION_UPDATE_WIDGET -> {
                updateWidget(intent.getIntExtra(EXTRA_WIDGET_ID, Int.MIN_VALUE))
            }
            else -> Log.wtf(TAG, "Unexpected action: ${intent.action}")
        }
    }

    private fun updateWidget(widgetId: Int) {
        val quote = loadQuote(this, widgetId) ?: return
        val views = RemoteViews(this.packageName, R.layout.stock_app_widget)
        views.setTextViewText(R.id.stock_symbol, quote.symbol)
        views.setTextViewText(R.id.bid_price, quote.bidPrice)
        views.setTextViewText(R.id.change, quote.change) // TODO: 11/10/2016 Support percentage
        AppWidgetManager.getInstance(this).updateAppWidget(widgetId, views)
    }
}

val ACTION_UPDATE_WIDGETS = "com.sam_chordas.android.stockhawk.service.action.UPDATE_WIDGETS"
val ACTION_UPDATE_WIDGET = "com.sam_chordas.android.stockhawk.service.action.UPDATE_WIDGET"
val EXTRA_WIDGET_IDS = "com.sam_chordas.android.stockhawk.service.extra.WIDGET_IDS"
val EXTRA_WIDGET_ID = "com.sam_chordas.android.stockhawk.service.extra.WIDGET_ID"

fun requestUpdateWidgets(context: Context, widgetIds: IntArray) {
    val intent = Intent(context, StockWidgetUpdateIntentService::class.java)
    intent.action = ACTION_UPDATE_WIDGETS
    intent.putExtra(EXTRA_WIDGET_IDS, widgetIds)
    context.startService(intent)
}

fun requestUpdateWidget(context: Context, widgetId: Int) {
    val intent = Intent(context, StockWidgetUpdateIntentService::class.java)
    intent.action = ACTION_UPDATE_WIDGET
    intent.putExtra(EXTRA_WIDGET_ID, widgetId)
    context.startService(intent)
}

private val PREFS_NAME = "com.sam_chordas.android.stockhawk.ui.StockAppWidget"
private val PREF_PREFIX_KEY = "appwidget_"

/**
 * Write the prefix to the SharedPreferences object for this widget
 */
fun saveStockSymbolPref(context: Context, appWidgetId: Int, text: String) {
    val prefs = context.getSharedPreferences(PREFS_NAME, 0).edit()
    prefs.putString(PREF_PREFIX_KEY + appWidgetId, text)
    prefs.apply()
}

/**
 * Read the prefix from the SharedPreferences object for this widget.
 * If there is no preference saved, get the default from a resource
 */
fun loadStockSymbolPref(context: Context, appWidgetId: Int): String? {
    val prefs = context.getSharedPreferences(PREFS_NAME, 0)
    return prefs.getString(PREF_PREFIX_KEY + appWidgetId, null)
}

fun loadQuote(context: Context, appWidgetId: Int): Quote? {
    val symbol = loadStockSymbolPref(context, appWidgetId) ?: return null
    val cursor = context.contentResolver.query(QuoteProvider.Quotes.CONTENT_URI,
            QUOTE_LIST_PROJECTION,
            QuoteColumns.ISCURRENT + " = ? AND " + QuoteColumns.SYMBOL + " = ?",
            arrayOf("1", symbol),
            null) ?: return null
    val quote: Quote?
    if (cursor.moveToFirst()) {
        val i = Quote.Indexes();
        i.set(cursor)
        quote = Quote()
        quote.set(cursor, i)
    } else quote = null
    cursor.close()
    return quote
}

fun deleteWidgetStockSymbolPref(context: Context, appWidgetId: Int) {
    val prefs = context.getSharedPreferences(PREFS_NAME, 0).edit()
    prefs.remove(PREF_PREFIX_KEY + appWidgetId)
    prefs.apply()
}
